client/hostvolumemanager/volume_fingerprint.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package hostvolumemanager
+
+import (
+	"context"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// this file is for fingerprinting *volumes*
+// *plugins* are detected in client/fingerprint/dynamic_host_volumes.go
+
+// HostVolumeNodeUpdater is used to add or remove volumes from the Node.
+type HostVolumeNodeUpdater func(name string, volume *structs.ClientHostVolumeConfig)
+
+// VolumeMap keys are volume `name`s, identical to Node.HostVolumes.
+type VolumeMap map[string]*structs.ClientHostVolumeConfig
+
+// UpdateVolumeMap returns true if it changes the provided `volumes` map.
+// If `vol` is nil, key `name` will be removed from the map, if present.
+// If it is not nil, `name: vol` will be set on the map, if different.
+//
+// Since it may mutate the map, the caller should make a copy
+// or acquire a lock as appropriate for their context.
+func UpdateVolumeMap(volumes VolumeMap, name string, vol *structs.ClientHostVolumeConfig) (changed bool) {
+	current, exists := volumes[name]
+	if vol == nil {
+		if exists {
+			delete(volumes, name)
+			changed = true
+		}
+	} else {
+		if !exists || !vol.Equal(current) {
+			volumes[name] = vol
+			changed = true
+		}
+	}
+	return changed
+}
+
+// WaitForFirstFingerprint implements client.FingerprintingPluginManager
+func (hvm *HostVolumeManager) WaitForFirstFingerprint(ctx context.Context) <-chan struct{} {
+	// the fingerprint manager puts batchFirstFingerprintsTimeout (50 seconds)
+	// on the context that it sends to us here so we don't need another
+	// timeout. we just need to cancel to report when we are done.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	volumes, err := hvm.restoreFromState(ctx)
+	if err != nil {
+		hvm.log.Error("failed to restore state", "error", err)
+		return ctx.Done()
+	}
+	for name, vol := range volumes {
+		hvm.updateNodeVols(name, vol) // => batchNodeUpdates.updateNodeFromHostVolume()
+	}
+	return ctx.Done()
+}
+func (hvm *HostVolumeManager) Run()      {}
+func (hvm *HostVolumeManager) Shutdown() {}
```

Comment:
The various RPC handlers pass a request context into the calls to the plugins (which is just a timeout on the background context), but we don't have a way of signalling to plugin calls that the client is shutting down to cancel them. There's two options:
* Stick an agent shutdown context on the `HostVolumeManager` that's closed here in `Shutdown`. And use a [joincontext](https://github.com/LK4D4/joincontext) as we do elsewhere in the client to handle this problem.
* Derive the request context in the RPC handler from an agent shutdown context. This is a much cleaner option... but I suspect the reason we don't have that elsewhere is maybe because the agent doesn't _have_ its own shutdown context? (Because the code base predates contexts.)

I don't think this is blocking to fix in this PR (once the agent process exits, those plugin calls are certainly going to stop!), but let's follow-up on that later.